### PR TITLE
MRG: anonymize_info function

### PR DESCRIPTION
**Pull request by kingjr**
We currently could only anonymize `raw` data. This allow anonymizing any `Info`, and add the method to `Epochs` so that we can share already preprocessed data that happened not to be anonymized.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Could we just add a method to the `ChannelsMixin`? Then all things that deal with channels will inherit it, including `Evoked`

**Comment by kingjr** [comment]
Sure

File: mne/io/meas_info.py [hunk 1]
```diff
@@ -267,6 +262,24 @@ def _update_redundant(self):
...
+        Measurement information for the dataset.
+    """
+    if not isinstance(info, Info):
+        raise ValueError('info must be an Info instance.')
+    if info.get('subject_info') is not None:
+        del info['subject_info']
+    info['meas_date'] = [0, 0]
+    for key_1 in ('file_id', 'meas_id'):
+        for key_2 in ('secs', 'msecs', 'usecs'):
+            info[key_1][key_2] = 0
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: also, why break this out into a function instead of just making the `info._anonymize` public?

**Comment by agramfort** [comment]
+1 for ChannelsMixin method although ChannelsMixin really starts to be a bad name for what it does... Maybe DataContainerMixin ?

**Comment by dengemann** [comment]
Very good idea. This will be useful.

File: mne/io/meas_info.py (hunk 1 above)
Comment (kingjr) [inline mne/io/meas_info.py]: Good point

**Comment by kingjr** [comment]
I don't find `class ChannelsMixin`. Do you mean `class SetChannelsMixin`?

**Comment by Eric89GXL** [comment]
Whatever mixin you think fits best :)

**Comment by kingjr** [comment]
I'm asking because I don't know these classes... :/

**Comment by Eric89GXL** [comment]
Yeah `SetChannelsMixin` should be fine I think, and as @agramfort mentioned renaming it would be good

**Comment by kingjr** [comment]
if we make `info.anonymize()` public, do we need to add the `anonymize` method in `ChannelsMixin` to have it accessible in `Epochs` and `Raw` etc? It seems a bit redundant, WDYT?

**Comment by dengemann** [comment]
I think the idea was not to really use info as abything but a dict
publicly. We had rejected several proposals to make info a real mne object.
But I agree it looks funny from a code perspective.

**Comment by Eric89GXL** [comment]
Hmm... maybe we shouldn't make `info._anonymize` public then, and instead have the Mixin have a `.anonymize` method that is a one-line wrapper

**Comment by kingjr** [comment]
> I think the idea was not to really use info as abything but a dict publicly. We had rejected several proposals to make info a real mne object. But I agree it looks funny from a code perspective.

Ok, it makes sense. I'm +1 on trying to keeping it as close to a dict so that we can ultimately clean this weird object.

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 2]
```diff
@@ -137,18 +137,6 @@ def test_subject_info():
     for key in keys:
         assert_equal(subject_info[key], raw_read.info['subject_info'][key])
     assert_equal(raw.info['meas_date'], raw_read.info['meas_date'])
-    raw.anonymize()
```
Comment (kingjr) [inline mne/io/fiff/tests/test_raw_fiff.py]: What follows is now saved in a specific `test_anonymize` function

File: mne/io/meas_info.py [hunk 3]
```diff
@@ -226,14 +226,6 @@ def __repr__(self):
         st %= non_empty
         return st
 
-    def _anonymize(self):
```
Comment (kingjr) [inline mne/io/meas_info.py]: To keep Info as close to a dict as possible, I'm removing this private method and call the public function `anonymize_info` instead.

**Comment by kingjr** [comment]
Ouch, that took me a while. `anonymize()` works fine, but when I save it into raw file, some fields of info (`info.['file_id']['secs']` and `info['meas_id']['secs']'`) are rewritten with different values than the anonymization.

It seems that `mne.io.write.write_id` necessarily writes some information independently of what is in `info`. I'm not sure whether this is full proof wrt anonymization because some labs may make an anonymize batch on the day of recording, in which case the date would still be in the fif. I'm not sure this is an issue, LMKWYT

Also we are currently not resetting some fields: e.g. `version`, `machid` and `usecs` remain intact in `meas_id` and `file_id`. Not sure whether this is a problem either.

File: mne/io/meas_info.py [hunk 4]
```diff
@@ -1513,3 +1505,32 @@ def _force_update_info(info_base, info_target):
...
             i_targ[key] = val
+
+
+def anonymize_info(info):
+    """Anonymize measurement information in place by removing 'subject_info',
+       'meas_date', 'file_id', 'meas_id' if they exist.
+
+    Parameters
+    ----------
+        info : dict, instance of Info
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: bad indentation

File: mne/io/meas_info.py [hunk 5]
```diff
@@ -1513,3 +1505,32 @@ def _force_update_info(info_base, info_target):
             continue
         for i_targ in info_target:
             i_targ[key] = val
+
+
+def anonymize_info(info):
+    """Anonymize measurement information in place by removing 'subject_info',
+       'meas_date', 'file_id', 'meas_id' if they exist.
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: Should be a single line, so you can put the actual fields below on a separate line if you want

File: mne/io/meas_info.py [hunk 6]
```diff
@@ -1513,3 +1505,32 @@ def _force_update_info(info_base, info_target):
...
+        Measurement information for the dataset.
+
+    Returns
+    -------
+        info : instance of Info
+        Measurement information for the dataset.
+
+    Notes
+    -----
+        Operates in place.
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: bad indentation

File: mne/io/meas_info.py [hunk 7]
```diff
@@ -1513,3 +1505,32 @@ def _force_update_info(info_base, info_target):
...
+       'meas_date', 'file_id', 'meas_id' if they exist.
+
+    Parameters
+    ----------
+        info : dict, instance of Info
+        Measurement information for the dataset.
+
+    Returns
+    -------
+        info : instance of Info
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: bad indentation

File: mne/io/meas_info.py [hunk 8]
```diff
@@ -1513,3 +1505,32 @@ def _force_update_info(info_base, info_target):
...
+        Measurement information for the dataset.
+
+    Notes
+    -----
+        Operates in place.
+    """
+    if not isinstance(info, (dict, Info)):
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: only check for `Info`

File: mne/io/tests/test_meas_info.py [hunk 9]
```diff
@@ -326,4 +327,44 @@ def test_check_consistency():
     assert_raises(RuntimeError, info2._check_consistency)
 
 
+def test_anonymize():
+    # Checks that contains sensitive information
```
Comment (jona-sassenhagen) [inline mne/io/tests/test_meas_info.py]: Syntax error :)

**Comment by kingjr** [comment]
Ready on my end

File: mne/channels/channels.py [hunk 10]
```diff
@@ -402,6 +403,25 @@ def plot_sensors(self, kind='topomap', ch_type=None, title=None,
...
 
+    def anonymize(self):
+        """Anonymize measurement information in place by removing
+        'subject_info', 'meas_date', 'file_id', 'meas_id' if they exist in
+        ``info``.
+
+        Returns
+        -------
+        inst : instance of Raw, Epochs or Evoked
+            The instance object.
```
Comment (agramfort) [inline mne/channels/channels.py]: Say you return self

**Comment by agramfort** [comment]
thanks @kingjr !

**Pull request merged**